quantum/process_keycode/process_tap_dance.c

```diff
@@ -158,6 +162,7 @@ bool process_tap_dance(uint16_t keycode, keyrecord_t *record) {
                 process_tap_dance_action_on_each_tap(action);
                 active_td = action->state.finished ? 0 : keycode;
             } else {
+                process_tap_dance_action_on_each_release(action);
                 if (action->state.finished) {
                     process_tap_dance_action_on_reset(action);
```

Comment:
I _suspect_ that the following change would be needed to fix the corner case when the `on_each_release` callback does `state->finished = true;` to finish the tap dance early (this kind of usage currently works for `on_each_tap` and is somewhat documented).
```suggestion
                    process_tap_dance_action_on_reset(action);
                    if (active_td == keycode) {
                        active_td = 0;
                    }
```
But I don't have time to test this just now (this would probably need another test to make sure that there are no superfluous `finished` calls after the tap dance timer expires).

And using `reset_tap_dance(state);` in `on_each_release()` would work properly anyway, because it resets `active_td` and clears the whole `state` (so the duplicate call of the `reset` callback won't happen); but it's not equivalent to `state->finished = true;` when used in the `on_each_tap` callback, so probably both cases should be supported.

Reply:
I just added your change and another test function that verifies the expected behavior.  Let me know if there are any other cases you think should be added.